### PR TITLE
Add test root readonly

### PR DESCRIPTION
**Pull request by sat0ken**
This implements the root_readonly_true validation in https://github.com/youki-dev/youki/issues/361

### CODE REVIEW

File: tests/contest/runtimetest/src/tests.rs
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:560]: I think this logic is wrong - we need to check if write access is false, not read-access is giving any error or not. Also if root is readonly , why would the read access give any error at all ? It should succeed, even ENOENT is not expected.

File: tests/contest/runtimetest/src/main.rs
Comment (YJDoc2) [inline tests/contest/runtimetest/src/main.rs:49]: Remove this extra blank

**Review by YJDoc2 — CHANGES_REQUESTED** [review]
Hey, I think we need to break this in two tests - one with readonly true and other with false. We will have a corresponding test in the `runtime` test, which will check that read operation is successful (maybe create a file and validate the contents) and the the write access will be tested according to config's value.

This is small deviation from the go test, but I think this would be a better test.

**Comment by sat0ken** [comment]
@YJDoc2 HI, I've fixed the code according to the comments. Could you please check the code?

File: tests/contest/runtimetest/src/main.rs
Comment (YJDoc2) [reply tests/contest/runtimetest/src/main.rs:49]: @sat0ken This is still to be done

File: tests/contest/contest/src/tests/root_readonly_true/root_readonly_tests.rs
Comment (YJDoc2) [inline tests/contest/contest/src/tests/root_readonly_true/root_readonly_tests.rs:26]: I think we should split this into two tests, maybe `root_readonly_true_test` and `root_readonly_false_test`. Current impl is correct, but I feel it is logically better to have two tests to check two behaviors rather than one test checking both.

File: tests/contest/runtimetest/src/tests.rs
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:559]: This is ok, but also check for read access is `Ok` and not err _if possible_. For the `!= EROFS` case, also add the actual `errno` in the `eprintl` for debugging

File: tests/contest/runtimetest/src/tests.rs
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:566]: I think for `readonly = false`, write access should not give any errors at all. So if it not ok, that should be considered error irrespective of errno.

**Review by YJDoc2 — CHANGES_REQUESTED** [review]
Hey, overall ok, but some changes needed.

File: tests/contest/runtimetest/src/main.rs
Comment (sat0ken) [reply tests/contest/runtimetest/src/main.rs]: Sorry, I fixed.
https://github.com/youki-dev/youki/pull/2976/commits/a49ca55f3b66d513820022b21b44b20371188ed7

File: tests/contest/contest/src/tests/root_readonly_true/root_readonly_tests.rs
Comment (sat0ken) [reply tests/contest/contest/src/tests/root_readonly_true/root_readonly_tests.rs]: @YJDoc2 
I split into tow test. Is this correct?
https://github.com/youki-dev/youki/pull/2976/commits/81d6128c54e1e8cf9cf92b182c6298be25fa6010

File: tests/contest/runtimetest/src/tests.rs
Comment (sat0ken) [reply tests/contest/runtimetest/src/tests.rs]: I fixed this.
https://github.com/youki-dev/youki/pull/2976/commits/c3c62fbf865aeaedfb6fdce1b33aba31301988ca

File: tests/contest/runtimetest/src/tests.rs
Comment (sat0ken) [reply tests/contest/runtimetest/src/tests.rs]: I fixed this.
https://github.com/youki-dev/youki/pull/2976/commits/c3c62fbf865aeaedfb6fdce1b33aba31301988ca

**Comment by sat0ken** [comment]
@YJDoc2 Thank you for review and comment! I fixed code, Could you please check the code?

**Comment by sat0ken** [comment]
@YJDoc2 Ping!

File: tests/contest/runtimetest/src/tests.rs
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:579]: Hey, for readonly=true, read_access should not give any error at all. So any error should be considered test fail. I think you might have incorrectly copy-pasted from the write access case. Please update.

File: tests/contest/runtimetest/src/tests.rs
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:590]: Why we are checking `raw_os_error().is_some()` ? If this was error, any error while writing, it should be considered as a test fail, right? Also can we have a read_access check in this branch as well? Basically :

if readonly=true{
   check write_access == error;
   check read_access == ok;

}else{
   check write_access == ok;
   check read_access == ok;
}

**Review by YJDoc2 — CHANGES_REQUESTED** [review]
some minor changes needed.

**Comment by sat0ken** [comment]
@YJDoc2 Thank you for review and comments. I've fixed the code. Could you please check the code?

File: tests/contest/runtimetest/src/tests.rs
Comment (YJDoc2) [inline tests/contest/runtimetest/src/tests.rs:578]:
```suggestion
        if let Err(e) = test_dir_read_access("/") {
            eprintln!(
                "readonly root filesystem, but error in testing read access for path /, error: {}",
                e
            );
    } else {
```
As I had mentioned in my previous comment,  we don't need specific check for raw_os_error. As long as this is error, the test is failing. Similar for the  two cases below this. Also I think here, you are converting the error to raw os error, and then using errorno to convert it back into the original error.

**Review by YJDoc2 — CHANGES_REQUESTED** [review]
One final change.

File: tests/contest/runtimetest/src/tests.rs
Comment (sat0ken) [reply tests/contest/runtimetest/src/tests.rs]: Sorry for any inconvenience caused by my misunderstanding.
I fixed this. Is this ok?

https://github.com/youki-dev/youki/pull/2976/commits/736127756158af9dd9328f8bb77cb08975f78858

Comment (YJDoc2) [reply tests/contest/runtimetest/src/tests.rs]: Yep, this is what I wanted :+1:

**Comment by YJDoc2** [comment]
@sat0ken , can you merge/rebase the latest main? the `test_inside_container` function has changed, so this will error if merged in current state. Apart from that this is good to go, so once that is done I'll merge.

Thanks for all the efforts, and apologies for so many back-and-forths :)

**Comment by sat0ken** [comment]
@YJDoc2 

> can you merge/rebase the latest main?
Sure ! I worked it. Is this ok?

https://github.com/youki-dev/youki/pull/2976/commits/397a8c5532f083d5dacf40a8a1c3b29fcfd62231

**Review by YJDoc2 — APPROVED** [review]
LGTM!
Thanks for your efforts, and apologies for so many back-and-forth :)

**Pull request merged**
